### PR TITLE
Implement route parameters

### PR DESCRIPTION
**Pull request by dmeijboom**
We really love the static-pages plugin but we are missing the ability to use route params. Without this we're forced to mix CMS pages (for detail-pages) with static-pages which isn't great because our clients can't edit pages that needs route parameters.

This pull request provides a working proof-of-concept and is able to render static-pages using route parameters (for example: /user/:id).

I would really like to hear if this would be accepted. If so I can finish the pull-request.

### CODE REVIEW

**Comment by LukeTowers** [comment]
Interesting, I didn't think that this would be this easy to accomplish. I'll talk to @daftspunk about it to see if he's fine with static pages having URL parameters in theory and then I'll let you know. Thanks for your work on this!

**Comment by dmeijboom** [comment]
Thanks! Any update on this?

**Comment by LukeTowers** [comment]
@dmeijboom is the idea to be able to use these parameters in snippets? How do you plan on achieving that?

**Comment by dmeijboom** [comment]
Exactly. I'm not sure but I'll look into this and update my pull-request this weekend.

**Comment by dmeijboom** [comment]
@LukeTowers Sorry for the delay. I fixed a bug where the filename was generated based on the URL parameters. In the static-page controller I now merge the parameters of the static-page into the parameters of the cms-page. This way they can be used in for instance components simply by using:

```twig
{{__SELF__.param('name')}}
```

Can you check this?

**Comment by LukeTowers** [comment]
@osmanzeki @w20k @munxar @alxy @bennothommo can you look into this and test?

**Comment by bennothommo** [comment]
@dmeijboom @LukeTowers I've had a bit of a look through. I will admit I could not get it working with a snippet derived from a partial - likely because static page contents are not run through the Twig parser, therefore it cannot feed the parameters through to the snippet inspector. I'll try again with a component-based snippet shortly.

It does not currently supported all types of route parameters - eg. optional route params or params that are limited by a regular expression. I've left a review for that.

**Comment by LukeTowers** [comment]
@bennothommo I don't think your review went through (at least I didn't see it show up yet)

**Comment by bennothommo** [comment]
@LukeTowers My bad - it was still pending.

**Comment by bennothommo** [comment]
@dmeijboom @LukeTowers Upon further testing, it does work with snippets from components, as @dmeijboom stated.

Personally, on an idea level, I prefer the separation of variable pages in the CMS section from "static" client-controlled pages (less avenues for the clients to ruin websites :P), but that's just my humble opinion.

File: classes/Page.php [hunk 1]
```diff
@@ -64,7 +64,7 @@ class Page extends ContentBase
      */
     public $rules = [
         'title' => 'required',
-        'url'   => ['required', 'regex:/^\/[a-z0-9\/_\-\.]*$/i', 'uniqueUrl']
+        'url'   => ['required', 'regex:/^\/[a-z0-9\/_\-\.\:]*$/i', 'uniqueUrl']
```
[the remark this answers is not in the document]
Comment (dmeijboom) [reply classes/Page.php]: You're right. Fixed.

**Comment by LukeTowers** [comment]
@dmeijboom any input on @bennothommo's comment about the separation?

**Comment by dmeijboom** [comment]
@bennothommo @LukeTowers 
> Personally, on an idea level, I prefer the separation of variable pages in the CMS section from "static" client-controlled pages (less avenues for the clients to ruin websites :P), but that's just my humble opinion.

Correct me if I'm wrong but are you suggesting not to implement route parameters for static-pages and only use them in CMS pages? If so this is very inconvenient for us since CMS pages can't be edited by our customers and there are a lot of use-cases for this (editable detail pages for instance). Even though the URL can be freely edited by the customer.

**Comment by bennothommo** [comment]
@dmeijboom What I'm saying is that I can't see the applicability of variable parameters in a static page. For example, if you were to set up a static page to view a product and have it load the product data based on the parameter from the URL, it implies that there is already a plugin somewhere where you can manage this data, so it is adding unnecessary complexity to then have the client also have to then create a static page and place a snippet somewhere on that page in order to see the data.

On the flip side, the site designer can simply set up the page in the CMS section, add a component with the necessary data, and have it just work, and then all the client has to do is manage the data through the plugin instead.

I'm more than happy to hear of examples of where this could be useful - I'm only basing this on work I've done previously. I also don't want to discourage you from bringing this forward - at the end of the day, this isn't my plugin. I was just suggesting that there might be better ways of implementing this that are already in October.

**Comment by dmeijboom** [comment]
@bennothommo I understand your point of view but I guess it really depends on the use-case. We use static-pages in combination with a repeater of blocks. A block is basically a partial or component for the frontend rendering and a form in the static-page plugin for the properties (such as title, content, image etc.). We developed blocks like: FAQ-block, Twitterfeed, Contactform etc. so our clients can build pages using these blocks. If we would use CMS pages for a product-detail page the layout can't be changed by our clients since these blocks are not available here.

@LukeTowers I'm not sure how this works but it would be really helpfull if someone decides to either accept or decline my pull-request now that both @bennothommo and I gave our feedback. If don't mind what the decision is but if it will be declined we can go forward with creating a fork. Thanks!

**Comment by maraerben** [comment]
Any progress here? This PR is exactly what we are looking for.

**Comment by alxy** [comment]
Does it impact performance in a significant way, especially when no routes parameters are used? Personally, I have to say I like the idea of this idea and see no reason to decline it when it doesn't harm performance for static pages significantly.

**Comment by LukeTowers** [comment]
@dmeijboom @maraerben @alxy all that's missing now is for people to test this PR and confirm that it works even while trying to break it and additions to the documentation to explain the new feature.

**Comment by bennothommo** [comment]
@dmeijboom @maraerben @alxy As per @LukeTowers message, I'm happy to merge this if you can thoroughly test it and provide the additions to the plugin documentation.

**Comment by bennothommo** [comment]
Closing as it has been over a month since any activity on this occurred and we are trying to figure out what issues are still relevant. If this is still something that you would like to see through to fruition please respond and we can get the ball rolling.

**Comment by MaTToX3** [comment]
So, any info on that? Or do you recommend that we use CMS Pages for such needs?

**Comment by LukeTowers** [comment]
@MaTToX3 we're not opposed to including this, we just need the following done first:

> I'm happy to merge this if you can thoroughly test it and provide the additions to the plugin documentation.

**Comment by dmeijboom** [comment]
I'm no longer working with the OctoberCMS so unfurtunately no time to create documentation and test it properly :/

**Comment by LukeTowers** [comment]
@dmeijboom thanks for your contributions on this so far, hopefully we can integrate it at some point!

**Comment by LukeTowers** [comment]
@maraerben @alxy either of you want to pick up the torch on this?

**Pull request closed without merging**